buffer/src/main/java/io/netty5/buffer/bytebuffer/NioBuffer.java

```diff
@@ -271,6 +271,42 @@ public void copyInto(int srcPos, Buffer dest, int destPos, int length) {
         InternalBufferUtils.copyToViaReverseLoop(this, srcPos, dest, destPos, length);
     }
 
+    @Override
+    public Buffer writeBytes(byte[] source, int srcPos, int length) {
+        if (source.length < srcPos + length) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+
+        checkWrite(woff, length, true);
+
+        if (hasWritableArray()) {
+            System.arraycopy(source, srcPos, writableArray(), writableArrayOffset(), length);
+        } else {
+            writableBuffer().put(source, srcPos, length);
+        }
+
+        skipWritableBytes(length);
+
+        return this;
+    }
+
+    @Override
+    public Buffer writeBytes(ByteBuffer source) {
+        final int length = source.remaining();
+
+        checkWrite(woff, length, true);
+
+        if (hasWritableArray()) {
+            source.get(writableArray(), writableArrayOffset(), length);
+        } else {
+            writableBuffer().put(source);
```

Comment:
`writableBuffer` allocates a slice. Try using `InternalBufferUtils.bbput`.

Reply:
Thanks for the insight, I was aware it created a duplicate but hadn't spotted anything that allowed me to avoid that.  I will check into that API and see what I can come up with.

Reply:
this has been changed to use bbput to attempt to avoid the ByteBuffer allocation.